### ISSUE TITLE
Make `delegate` to be a macro implemented by the compiler

### ISSUE THREAD

**Issue by Sija** [crystal-lang/crystal#9927]
I'd actually prefer if we change `delegate` to be a macro implemented by the compiler. Then whenever we find a call to that delegated method, we replace it with the delegation. It will solve this problem and others (like capturing block arguments not possible right now), even though it's a "magic" solution.

_Originally posted by @asterite in https://github.com/crystal-lang/crystal/issues/9683#issuecomment-727553273_

**Comment by BrucePerens** [comment]
I was able to get pretty far with making a better `delegate`, it is type-safe and copies the argument list from the delegated-to method, and it handles blocks. But the syntax is slightly different. This version requires that you provide a class name, so `A.delegate2 method-name, to: instance-of-A`.
```crystal
class Object
  macro delegate2(name, to)
    {% methods = @type.methods %}
    {% got_name = false %}
    {% for m in methods %}
      {% if m.name.id == name.id %}
        {% got_name = true %}
        {% has_yield = false %}
        {% if m.body.class_name == "Expressions" %}
          {% for node in m.body.expressions %}
            {% if node.class_name == "Yield" %}
              {% has_yield = true %}
            {% end %}
          {% end %}
        {% end %}
        {% if m.body.class_name == "Yield" %}
          {% has_yield = true %}
        {% end %}
        def {{name}}({{m.args.join(",").id}}) {{ m.return_type.stringify != "" ? ": #{m.return_type}".id : "".id }}
          {{to}}.{{name}}({{m.args.map{ |arg| arg.internal_name }.join(", ").id}}) {{has_yield ? "do |*args| yield *args end".id : "".id}}
        end
      {% end %}
    {% end %}
    {% if !got_name %}
      {% raise "delegate: Method \"#{name}\" wasn't found in \"#{@type.name}\"." %}
    {% end %}
    {% debug %}
  end
end

class A
  def foo(a : Int32) : String
    yield
  end
end

class B
  @a = A.new
  A.delegate2(foo, to: @a)
end

b = B.new
p b.foo(1) { "Hello" }
```

**Comment by BrucePerens** [comment]
If someone could show me how to get rid of the `A` in `A.delegate2 foo, to: @a`, the above would solve all the issues that I think @sija has.

**Comment by asterite** [comment]
There's no way to do it because when the macro runs type information for the instance variable, or whatever expression you pass, is not available.

Also, the current delegate macro works even before the method is declared on the target type. Plus it works if the target is a union. The delegate macro can really be best implemented by having the compiler implement it.

**Comment by BrucePerens** [comment]
@asterite Wouldn't this work if the macro language included a `typenode()` operator, which would return a typenode referring to the argument? So that one could use `typenode(@a).methods`? I guess it would have to synthesize a typenode for unions.

**Comment by asterite** [comment]
When the macro is executed the type information of instance variables isn't yet computed. So typenode(@a) won't work because there's no type yet at that moment.

**Comment by BrucePerens** [comment]
OK, so the only way for this to work for all cases is for it to be a compiler keyword, and for generation of the delegator method to be deferred until the delegated-to method is defined.

**Comment by asterite** [comment]
I think so. It could also be implemented by method missing. The problem is that there's only one method_missing. Maybe it would be nice to be able to register handlers for method_missing at the language level...

**Comment by asterite** [comment]
Actually, we could make each method_missing be able to call the previous definition if it exists, so you can form a chain. I might play with this idea...

**Comment by BrucePerens** [comment]
Please code any new `delegate` keyword so that it is possible to separately name the delegator method. I have a use case of extending `delegate` to wrap and unwrap arguments and return values, in https://github.com/BrucePerens/recursive_generic . At least unless/until recursively-defined aliases get fixed.

**Comment by asterite** [comment]
Well, it seems there's no way to capture a block with `method_missing`.

I personally think we should remove `delegate`. It's very easy (though admittedly tedious the first time) to define the actual method that delegates the call. And once you write it there's no need to touch that code again. Well, I guess if the underlying method changes... but that's an easy change too.

**Comment by BrucePerens** [comment]
It's not so bad to insist on the user providing the type information, then, as I currently do in `delegate2`. Even if the user has to find the union member providing a method, and if the method has to already be defined. 

@asterite I've noticed you're eager to remove *many* language features. And, I guess, can't convince the team.

I ran into another issue, which is that my `delegate2` must delegate *all* methods by that name in the target.

**Comment by asterite** [comment]
Yes, I'd like to remove things that don't work 100% of the time. They can be added later when we find a way to make them work in all cases.

That said, I use `delegate` a lot in the compiler and it always worked fine.
